xayn-ai-ffi-c/src/reranker/ai.rs

```diff
@@ -316,6 +339,58 @@ pub unsafe extern "C" fn xaynai_analytics(
 #[no_mangle]
 pub unsafe extern "C" fn xaynai_drop(_xaynai: Option<Box<CXaynAi>>) {}
 
+/// Serializes the synchronizable data of the reranker.
+///
+/// # Errors
+/// Returns a null pointer if:
+/// - The xaynai is null.
+/// - The serialization fails.
+/// - An unexpected panic happened.
+///
+/// # Safety
+/// The behavior is undefined if:
+/// - A non-null xaynai doesn't point to memory allocated by [`xaynai_new()`].
+/// - A non-null error doesn't point to an aligned, contiguous area of memory with a [`CError`].
+#[no_mangle]
+pub unsafe extern "C" fn xaynai_syncdata_bytes(
+    xaynai: Option<&CXaynAi>,
+    error: Option<&mut CError>,
+) -> Option<Box<CBytes>> {
+    let sync_bytes = AssertUnwindSafe(
+        // Safety: The mutable memory is not accessed in this call.
+        || unsafe { CXaynAi::syncdata_bytes(xaynai) },
+    );
+
+    call_with_result(sync_bytes, error)
+}
+
+/// Synchronizes the internal data of the reranker with another.
+///
+/// # Errors
+/// - The `xaynai` is null.
+/// - The serialized data `bytes` is invalid.
+/// - The synchronization failed.
+/// - An unexpected panic happened.
+///
+/// In case of a [`CCode::Panic`] the Xayn AI must not be accessed anymore and should be dropped via
+/// [`xaynai_drop()`]. The last known valid state can be restored by the caller via [`xaynai_new()`]
+/// with a previously serialized reranker database obtained from [`xaynai_serialize()`].
+///
+/// # Safety
+/// The behavior is undefined if:
+/// - A non-null `xaynai` doesn't point to memory allocated by [`xaynai_new()`].
+/// - A non-null `error` doesn't point to an aligned, contiguous area of memory with a [`CError`].
+#[no_mangle]
+pub unsafe extern "C" fn xaynai_synchronize(
+    xaynai: Option<&mut CXaynAi>,
+    bytes: CBytes,
```

Comment:
usually fat pointers won't work for the dart wrapper. since we get a thin pointer from `xaynai_syncdata_bytes()` anyways, i'd suggest to use it as input for this method too.
```suggestion
    bytes: Option<Box<CBytes>>,
```

alternatively this function could accept non-owned bytes, which probably makes more sense as this is likely to be called on a different device than the bytes getter.
```suggestion
    bytes: Option<&CBytes>,
```

Reply:
thanks! I've gone with the latter suggestion.